### DISCUSSION TITLE
KEYPAD for compatibility

### DISCUSSION THREAD

**Discussion by SF73** [ArminJo/DigistumpArduino#36] [Q&A]
Hi,
I would like to maximize the compatibility between different layouts and I noticed that sending
```DigiKeyboard.sendKeyStroke(KEYPAD_X) ``` seems to work whatever the layout.
Is there possible to handle strings in `DigiKeyboard.println` the same way (i.e DigiKeyboard.println("1234") typing1234 whatever the layout)?

Thanks !

**Comment by ArminJo** [comment]
As long as this characters are layout independent you will have success 😀

**Comment by SF73** [comment]
`KEYPAD_X` are probably layout independent but when I use `DigiKeyboard.println("123")` each char seems to be mapped not to KEYPAD but to numbers above the AZERTY/QWERTY part.
`DigiKeyboard.println("123")`  basically outputs `123` or `&é"` depending on the layout but `DigiKeyboard.sendKeyStroke(KEYPAD_1)` always produces a `1`. I would like to "patch" `println` to have the same property.

**Reply by ArminJo** [reply]
> outputs `123` or `&é"` depending on the layout

The layout must match the one in windows, so you have two keybords installed in windows and test with the window layout you use in the digikeyboard?

**Reply by SF73** [reply]
Yes exactly. The idea is that I would like to use the same digispark on different keyboard layouts.
To do so, I thought I could simply write letters by sending their `alt+XXX` code but this is quite slow as it requires multiple ```sendKeyPress``` for a single char.

```c
void sendCode(int n) {
  unsigned int NUMKEY[] = {KEYPAD_0, KEYPAD_1, KEYPAD_2, KEYPAD_3, KEYPAD_4, KEYPAD_5, KEYPAD_6, KEYPAD_7, KEYPAD_8, KEYPAD_9};
  int ones = n % 10;
  int tens = (n / 10) % 10;
  int hund = (n / 100) % 10;
  DigiKeyboard.sendKeyPress(NUMKEY[hund], MOD_ALT_LEFT);
  DigiKeyboard.sendKeyPress(0, MOD_ALT_LEFT);
  DigiKeyboard.sendKeyPress(NUMKEY[tens], MOD_ALT_LEFT);
  DigiKeyboard.sendKeyPress(0, MOD_ALT_LEFT);
  DigiKeyboard.sendKeyPress(NUMKEY[ones], MOD_ALT_LEFT);
  DigiKeyboard.sendKeyPress(0, 0);
}
```
With this function, `sendCode(97)` would produce an `a` no matter the layout.

So as it is not ideal, I like to speed up this by first replacing 0-9 in `println` with their KEYPAD counterpart.
